Kernel implementation

KevinHu-Create opened the pull request:
Summary

- Implemented `InventoryManager1L` as the kernel implementation for InventoryManager component
- Used `Map<String, Integer>` as internal representation for inventory state
- Implemented all kernel methods:
  - `addItem`, `removeItem`, `hasItem`, `getItemCount`, `getAllItems`
- Implemented all standard methods:
  - `newInstance`, `clear`, `transferFrom`
- All methods follow design-by-contract

KevinHu-Create committed:
Implemented InventoryManager and fixed all errors

KevinHu-Create committed:
complete CHANGELOG.md and complete the question about why choose this component to implement

KevinHu-Create committed:
Added InventoryManagerInterface and updated InventoryManager to implement it and updated the CHANGELOG.md

KevinHu-Create committed:
Added Inventory Manager Hierarchy Diagram

KevinHu-Create committed:
update the Hierachy graph

KevinHu-Create committed:
Completed InventoryManager component with kernel and enhanced interfaces

KevinHu-Create committed:
Implemented InventoryManagerSecondary and secondary methods

- Added clearInventory, isStockLow, and restock
- Used kernel methods only
- Added getAllItems to kernel
- Added OSU CSE components. jar

KevinHu-Create committed:
modified Impl.java

This class should implement all method in the interface including standard, now here added all the implementation to these methods

KevinHu-Create committed:
Delete newInstance() in InventoryManager interface

It is inherited from InventoryManagerKernel, no need to override it again.

KevinHu-Create committed:
Updated CHANGELOG.md and the part4 .md files

KevinHu-Create committed:
Completed Part 5: Kernel Implementation of InventoryManager

Summary

- Implemented `InventoryManager1L` as the kernel implementation for InventoryManager component
- Used `Map<String, Integer>` as internal representation for inventory state
- Implemented all kernel methods:
  - `addItem`, `removeItem`, `hasItem`, `getItemCount`, `getAllItems`
- Implemented all standard methods:
  - `newInstance`, `clear`, `transferFrom`
- All methods follow design-by-contract

jrg94 commented on src/InventoryManager1L.java:
```diff
@@ -0,0 +1,179 @@
+     * @ensures this = #source and source = empty
+     */
+    @Override
+    public void transferFrom(InventoryManager source) {
```
These methods are all inherited, so you shouldn't need to document them.

jrg94 commented on src/InventoryManager1L.java:
```diff
@@ -0,0 +1,179 @@
+ */
+public class InventoryManager1L extends InventoryManagerSecondary {
+
+    private final Map<String, Integer> inventory;
```
You should document this.

jrg94 commented on src/InventoryManager1L.java:
```diff
@@ -0,0 +1,179 @@
+     */
+    @Override
+    public void clear() {
+        this.clearInventory(); // calls the secondary method from superclass
```
For your standard methods, i would recommend looking at old projects for how to implement these. Most of them are implemented in the same way using createNewRep or other techniques.

KevinHu-Create committed:
Complete all method inherited from standard

Follow the instructions that use createNewRep to implement standard method.

Get clear that I should override all standard method in InventoryManager1L even I may not use them because of the interface contract

When we override equals we should also override Hashcode().

jrg94 commented on src/InventoryManager1L.java:
```diff
@@ -0,0 +1,210 @@
+ */
+public class InventoryManager1L extends InventoryManagerSecondary {
+
+    private  Map<String, Integer> inventory;
```
Make sure you document this.

jrg94 commented on src/InventoryManager1L.java:
```diff
@@ -0,0 +1,210 @@
+        this.inventory = this.createNewRep();
+    }
+
+    @Override
```
These common methods should only be in secondary, not in the kernel implementation.

jrg94 replied:
These haven't been removed like needed.

jrg94 commented on src/InventoryManager1L.java:
```diff
@@ -0,0 +1,210 @@
+     */
+    @Override
+    public void clear() {
+        this.inventory = this.createNewRep();
```
Usually, createNewRep is a void method. It would be called to effectively clear the data structure.

jrg94 commented on src/InventoryManager1L.java:
```diff
@@ -0,0 +1,210 @@
+     */
+    @Override
+    public InventoryManager newInstance() {
+        return new InventoryManager1L();
```
This isn't implemented quite how I would expect. See old projects, but there's a try/catch (I think) needed.

jrg94 commented on src/InventoryManager1L.java:
```diff
@@ -0,0 +1,210 @@
+     *
+     * @return a new empty map
+     */
+    private Map<String, Integer> createNewRep() {
```
```suggestion
    private void createNewRep() {
```

jrg94 commented on src/InventoryManager1L.java:
```diff
@@ -0,0 +1,210 @@
+     * @return a new empty map
+     */
+    private Map<String, Integer> createNewRep() {
+        return new HashMap<>();
```
Set `inventory` directly here.

jrg94 commented on src/InventoryManager1L.java:
```diff
@@ -0,0 +1,210 @@
+
+        this.clearInventory();
+
+        for (String item : source.getAllItems()) {
```
transferFrom is an O(1) method. You can swap your reps directly.

jrg94 reviewed (changes requested):
Similar issues as before. Hopefully, these comments help more.

KevinHu-Create committed:
Changes follow professor's instructions

Implemented kernel transferFrom with O(1) rep swap
Removed standard transferFrom from Secondary
Refactored createNewRep() to follow OSU convention
